File: 01-programming/python/concepts/code/01-basics/simple-strings.py
# -o-o-o-o- Simple Strings -o-o-o-o-

# This file goes over basic strings in Python simply.
#
#  # # # # KEY # # # #
#
# str      - Python's built-in string type. Even simpler than C#.
#            In C, strings are char arrays: char str[50];
#            In C#, strings are objects:    string str = "Hello";
#            In Python, strings are just:  str = "Hello"   or  str = 'Hello'
#            Single and double quotes both work in Python.
#
# f"..."   - f-string (formatted string). Embeds variables directly.
#            eg. f"Hello, {name}!"
#            Equivalent to $"Hello, {name}!" in C# or printf in C.
#
# +        - string concatenation. Joins two strings together. Same as C#.
#
# len()    - returns the number of characters in a string.
#            Equivalent to .Length in C#.
#
# Strings are immutable in Python — just like in C#. Methods return new strings.
#
# .upper()      - returns the string in UPPERCASE.
# .lower()      - returns the string in lowercase.
# .strip()      - removes whitespace from both ends (like .Trim() in C#).
# .replace()    - replaces all occurrences of a substring with another.
# .split()      - splits a string into a list based on a separator.
# .join()       - joins a list of strings into one string with a separator.
# .find()       - returns the index of the first occurrence (-1 if not found).
# .startswith() - returns True if the string starts with a given prefix.
# .endswith()   - returns True if the string ends with a given suffix.
# in            - checks if a substring exists in a string (returns True/False).
#                 Equivalent to .Contains() in C#.
#
# input()  - reads a line of user input as a string.
# print()  - prints text to the console with a newline at the end.
#

# Declaring strings
first_name = "John"
last_name  = 'Doe'   # single quotes work the same as double quotes

# Concatenation
full_name = first_name + " " + last_name
print("Full name: " + full_name)

# f-string interpolation (preferred in Python)
print(f"Hello, {first_name} {last_name}!")

# String length
print(f"\nLength of '{full_name}': {len(full_name)}")

# Case methods
print(f"Uppercase: {full_name.upper()}")
print(f"Lowercase: {full_name.lower()}")

# Checking contents
print(f"\nContains 'John': {'John' in full_name}")   # True (uses 'in' keyword)
print(f"Starts with 'Jo': {full_name.startswith('Jo')}")
print(f"Ends with 'oe':   {full_name.endswith('oe')}")

# Replacing text
replaced = full_name.replace("John", "Jane")
print(f"\nAfter replace: {replaced}")

# Finding a character's index
index = full_name.find("D")
print(f"Index of 'D': {index}")

# Slicing a string (like Substring in C#)
# string[start:end]  — end is not included
sub = full_name[0:4]  # gets "John"
print(f"Slice [0:4]: {sub}")

# Splitting a string into a list
csv = "apple,banana,cherry"
fruits = csv.split(",")
print(f"\nSplit result: {fruits}")

# Joining a list back into a string
joined = " - ".join(fruits)
print(f"Joined: {joined}")

# Multi-line strings (using triple quotes)
multi = """This is a
multi-line
string."""
print(f"\nMulti-line string:\n{multi}")

# Getting user input (always returns a string)
user_input = input("\nEnter your name: ")
print(f"Hello, {user_input.strip()}!")   # strip() removes extra whitespace

File: 01-programming/csharp/concepts/code/01-basics/simple-strings.cs
// -o-o-o-o- Simple Strings -o-o-o-o-

/* This file goes over basic strings in C# simply.
 *
 *  # # # # KEY # # # #
 *
 * string  - a built-in type in C# for storing text. Far easier than C.
 *           In C, strings are char arrays with a null terminator: char str[50];
 *           In C#, string is a proper type: string name = "Hello";
 *           Strings in C# are immutable (cannot be changed after creation).
 *
 * $"..."  - string interpolation. Embed variables directly into a string.
 *           eg. $"Hello, {name}!" is cleaner than "Hello, " + name + "!"
 *
 * +       - string concatenation. Joins two strings together.
 *           eg. "Hello" + " World" = "Hello World"
 *
 * .Length - property that returns the number of characters in a string.
 *           eg. "Hello".Length returns 5
 *
 * .ToUpper()    - returns the string in ALL UPPERCASE.
 * .ToLower()    - returns the string in all lowercase.
 * .Trim()       - removes whitespace from the start and end of a string.
 * .Contains()   - returns true if a string contains a given substring.
 * .Replace()    - replaces all occurrences of one substring with another.
 * .Substring()  - extracts part of a string starting at a given index.
 * .Split()      - splits a string into an array based on a separator character.
 * .IndexOf()    - returns the index of the first occurrence of a character/string.
 *
 * Console.WriteLine() - prints text to the console with a newline at the end.
 * Console.ReadLine()  - reads a line of user input as a string.
 *
 * */

using System;

class Program {
    static void Main() {

        // Declaring and initialising strings
        string firstName = "John";
        string lastName  = "Doe";

        // Concatenation
        string fullName = firstName + " " + lastName;
        Console.WriteLine("Full name: " + fullName);

        // String interpolation (preferred in C#)
        Console.WriteLine($"Hello, {firstName} {lastName}!");

        // String properties and methods
        Console.WriteLine($"\nLength of '{fullName}': {fullName.Length}");
        Console.WriteLine($"Uppercase: {fullName.ToUpper()}");
        Console.WriteLine($"Lowercase: {fullName.ToLower()}");

        // Checking contents
        bool hasJohn = fullName.Contains("John");
        Console.WriteLine($"Contains 'John': {hasJohn}");

        // Replacing text
        string replaced = fullName.Replace("John", "Jane");
        Console.WriteLine($"After Replace: {replaced}");

        // Extracting a substring (start index, length)
        string sub = fullName.Substring(0, 4); // gets "John"
        Console.WriteLine($"Substring (0, 4): {sub}");

        // Finding a character's index
        int index = fullName.IndexOf('D');
        Console.WriteLine($"Index of 'D': {index}");

        // Splitting a string into parts
        string csv = "apple,banana,cherry";
        string[] fruits = csv.Split(',');
        Console.WriteLine("\nSplit result:");
        foreach (string fruit in fruits) {
            Console.WriteLine($"  {fruit}");
        }

        // Getting user input (always returns a string)
        Console.Write("\nEnter your name: ");
        string userInput = Console.ReadLine();
        Console.WriteLine($"Hello, {userInput.Trim()}!"); // Trim removes extra spaces

    }
}
